src/Microsoft.Tye.Hosting/Model/Application.cs

```diff
@@ -80,78 +83,80 @@ string GetValueFromBinding(EnvironmentVariableSource source)
-                //
-                // - connectionString from config
-                // - URI
-                // - host:port
-                //
-                // This is used by the GetConnectionString method, commonly when the service providing the
-                // binding isn't the user's code.
-                if (!string.IsNullOrEmpty(b.ConnectionString))
+                if (!string.IsNullOrEmpty(binding.ConnectionString))
                 {
                     // Special case for connection strings
-                    set($"CONNECTIONSTRINGS__{configName}", b.ConnectionString);
-                }
-                else if (!string.IsNullOrEmpty(b.Protocol) && b.Port != null)
-                {
-                    set($"CONNECTIONSTRINGS__{configName}", $"{b.Protocol}://{host}:{b.Port}");
-                }
-                else if (b.Port != null)
-                {
-                    set($"CONNECTIONSTRINGS__{configName}", $"{host}:{b.Port}");
+                    var connectionString = TokenReplacement.ReplaceValues(binding.ConnectionString, binding, bindings);
+                    set($"CONNECTIONSTRINGS__{configName}", connectionString);
+                    return;
                 }
 
-                if (!string.IsNullOrEmpty(b.Protocol))
+                if (!string.IsNullOrEmpty(binding.Protocol))
                 {
                     // IConfiguration specific (double underscore ends up telling the configuration provider to use it as a separator)
-                    set($"SERVICE__{configName}__PROTOCOL", b.Protocol);
-                    set($"{envName}_SERVICE_PROTOCOL", b.Protocol);
+                    set($"SERVICE__{configName}__PROTOCOL", binding.Protocol);
+                    set($"{envName}_SERVICE_PROTOCOL", binding.Protocol);
                 }
 
-                if (b.Port != null)
+                if (binding.Port != null)
                 {
-                    var port = (service.Description.RunInfo is DockerRunInfo) ? b.ContainerPort ?? b.Port.Value : b.Port.Value;
-
-                    set($"SERVICE__{configName}__PORT", port.ToString());
-                    set($"{envName}_SERVICE_PORT", port.ToString());
+                    set($"SERVICE__{configName}__PORT", binding.Port.Value.ToString(CultureInfo.InvariantCulture));
+                    set($"{envName}_SERVICE_PORT", binding.Port.Value.ToString(CultureInfo.InvariantCulture));
                 }
 
-                set($"SERVICE__{configName}__HOST", host);
-                set($"{envName}_SERVICE_HOST", host);
+                set($"SERVICE__{configName}__HOST", binding.Host);
+                set($"{envName}_SERVICE_HOST", binding.Host);
             }
+        }
+
+        // Compute the list of bindings visible to `service`. This is contextual because details like
+        // whether `service` is a container or not will change the result.
+        private List<EffectiveBinding> ComputeBindings(Service service, string defaultHost)
```

Comment:
not a fan of local functions yet?

Reply:
This could be a local function but it would close over `this` and so couldn't be static. I wasn't sure how the compiler represents that I just used private.